### ISSUE TITLE
Missing custom event for relation removal

### ISSUE THREAD

**Issue by sed-i** [canonical/tls-certificates-interface#25]
I'm integrating the `tls-certificates-interface` in traefik: https://github.com/canonical/traefik-k8s-operator/pull/100.
When I `juju remove-relation` between `tls-certificates-operator` and `traefik`, the only way to remove the self-signed cert (that came from `tls-certificates-operator` over relation data) is by manually observing the "relation-broken" event.

Typically we say that a lib is responsible for registering the relation events on behalf of the charm, and emit custom events per need.

Currently the requirer side has these events:

https://github.com/canonical/tls-certificates-interface/blob/f50ed0231f6eb5a6650057c6b6fcb1ae22cae73d/lib/charms/tls_certificates_interface/v1/tls_certificates.py#L840-L846

none of which is emitted on relation broken.

Ideally I would like to observe a custom event instead of "breaking the pattern" and observe "tls-certificates" relation events directly. Wdyt?

**Comment by gruyaume** [comment]
Good point, I'm not sure however that we should re-use one of those events in the case of a relation broken. For instance, the broken relation doesn't necessarily mean that the certificate was either revoked or expired. We could wrap the `relation-broken` event in something like `certificate-relation-removed` but I'm not sure what this would add in terms of value. What do you think?

**Comment by ghislainbourgeois** [comment]
I think this could be valuable to alert a human operator to the fact that this relation is broken. I think most charms would want to keep using the certificate until it expires, but having a broken relation means that the certificate will not automatically renew, and that could create an outage.

For that reason, I think a custom event like `CertificateProviderGoneEvent` (or a better name) could be added.

**Comment by sed-i** [comment]
We already have a lot of events to consider.

I think that semantically, a removed relation is equivalent to revoked.

Moreover, afaict, `expired` and `revoked` can be consolidated into `certificate_no_longer_valid(reason=expired|revoked)`.

**Comment by gruyaume** [comment]
I agree with the two following statements:
- Expired and revoked can be consolidated into certificate_no_longer_valid(reason=expired|revoked).
- A custom event like CertificateRelationRemovedEvent could be added.

However I don't agree that `a removed relation is equivalent to revoked`. The word `revoked` has some specific meaning attached to it and I see no reason why we should change this meaning. For instance, it is possible for the relation to be removed and the certificate to still be valid and it is also possible for the relation to be intact and the certificate to be revoked.

**Comment by sed-i** [comment]
I agree that `revoked` has a certain meaning.

The only scenario I can think of for when the lifetime of a certificate is longer than the lifetime of the relation is debugging charms in error state. And even then one could argue that when re-related, the charm could send the same CSR and get the same cert back.

With juju, relation data cannot outlive the relation, and likewise a cert that was sent over relation data cannot outlive the relation. This is consistent with the model. Keeping the cert after the relation is gone is inconsistent with the juju model.

It seems natural to me to consolidate `revoked`, `expired` and `relation-broken` into `certificate_no_longer_valid(reason)`: the charm on the other side of the relation has nothing to do but to dispose of the cert (and perhaps issue a follow-up action based on the `reason`).

**Comment by gruyaume** [comment]
I agree and I'm fine with placing the relation-broken part of the reasons for the`certificate_no_longer_valid` event.

**Comment by ghislainbourgeois** [comment]
I also agree with this design. I would propose `CertificateInvalidatedEvent` for the name of the event.
